src/sage/combinat/bijectionist.py

```diff
@@ -0,0 +1,3217 @@
+            {[]: 0, [1]: 1, [1, 2]: 0, [2, 1]: 2, [1, 2, 3]: 1, [1, 3, 2]: 1, [2, 1, 3]: 0, [2, 3, 1]: 0, [3, 1, 2]: 3, [3, 2, 1]: 1}
+            {[]: 0, [1]: 1, [1, 2]: 0, [2, 1]: 2, [1, 2, 3]: 1, [1, 3, 2]: 1, [2, 1, 3]: 1, [2, 3, 1]: 0, [3, 1, 2]: 3, [3, 2, 1]: 0}
+
+        Calling this with non-equidistributed statistics yields an error::
+
+            sage: bij = Bijectionist(A, B, fix)
+            sage: bij.set_statistics((wex, fix))
+            Traceback (most recent call last):
+            ...
+            ValueError: Statistics alpha and beta are not equidistributed!
+
+        TESTS:
+
+        Calling ``set_statistics`` without arguments should restore the previous state::
+
+            sage: N = 3; A = B = [permutation for n in range(N) for permutation in Permutations(n)]
+            sage: def wex(p): return len(p.weak_excedences())
+            sage: def fix(p): return len(p.fixed_points())
+            sage: def des(p): return len(p.descents(final_descent=True)) if p else 0
+            sage: bij = Bijectionist(A, B, fix)
+            sage: bij.set_statistics((wex, des), (len, len))
+            sage: for solution in bij.solutions_iterator():
+            ....:     print(solution)
+            {[]: 0, [1]: 1, [1, 2]: 0, [2, 1]: 2}
+            sage: bij.set_statistics()
+            sage: for solution in sorted(list(bij.solutions_iterator()), key=lambda d: tuple(sorted(d.items()))):
+            ....:     print(solution)
+            {[]: 0, [1]: 0, [1, 2]: 1, [2, 1]: 2}
+            {[]: 0, [1]: 0, [1, 2]: 2, [2, 1]: 1}
+            {[]: 0, [1]: 1, [1, 2]: 0, [2, 1]: 2}
+            {[]: 0, [1]: 1, [1, 2]: 2, [2, 1]: 0}
+            {[]: 0, [1]: 2, [1, 2]: 0, [2, 1]: 1}
+            {[]: 0, [1]: 2, [1, 2]: 1, [2, 1]: 0}
+            {[]: 1, [1]: 0, [1, 2]: 0, [2, 1]: 2}
+            {[]: 1, [1]: 0, [1, 2]: 2, [2, 1]: 0}
+            {[]: 1, [1]: 2, [1, 2]: 0, [2, 1]: 0}
+            {[]: 2, [1]: 0, [1, 2]: 0, [2, 1]: 1}
+            {[]: 2, [1]: 0, [1, 2]: 1, [2, 1]: 0}
+            {[]: 2, [1]: 1, [1, 2]: 0, [2, 1]: 0}
+
+        """
+        self._bmilp = None
+        self._n_statistics = len(alpha_beta)
+        # TODO: do we really want to recompute statistics every time?
+        self._alpha = lambda p: tuple(arg[0](p) for arg in alpha_beta)
+        self._beta = lambda p: tuple(arg[1](p) for arg in alpha_beta)
+
+        # generate fibers
+        self._statistics_fibers = {}
+        for a in self._A:
+            v = self._alpha(a)
+            if v not in self._statistics_fibers:
+                self._statistics_fibers[v] = ([], [])
+            self._statistics_fibers[v][0].append(a)
+
+        for b in self._B:
+            v = self._beta(b)
+            if v not in self._statistics_fibers:
+                raise ValueError(f"Statistics alpha and beta do not have the same image, {v} is not a value of alpha, but of beta!")
```

Comment:
I think we should follow Python conventions with the error messages and start with a lower case and not end with punctuation. (Sage generally follows this.)